Refactor/question : 리팩토링 및 테스트 코드 검증

Kimseungin0529 opened the pull request.

Commit:
refactor : QuestionContent enum 최적화

enum 수정으로 인한 Question, QuestionServiceImp, QuestionServiceImpTest 변경

Commit:
refactor : Questions 일급 컬렉션 구현 및 일부 수정

Commit:
refactor : QuestionService, QuestionServiceImp 네이밍 변경 및 @transactional 삭제

. DB 접근 로직이 없고 단순히 Question에 대한 공급만 하기에 Service 보다는 Provider의 의미가 더 맞아 보여서 수정
2. DB 관련 로직이 없기에 @transactional의 의미가 없고 오히려 있는 것이 읽는이에게 혼동을 줄 수 있다고 판단하여 수정
(Question 객체(엔티티)는 Analysis 에 종속적이라 Analysis 부분에서 저장, 삭제 등 관리)

3. QuestionServiceImpTest 부분에서 검증한 것은 QuestionService가 아닌 QuestionContent enum에 대한 검증이라 테스트 클래스명 및 패키지 경로 수정

Commit:
test : test : QuestionProvidable.createRandomQuestions 메서드 테스트 검증

Commit:
chore : m1 관련 netty 실행 오류로 인한 의존성 주입

Commit:
style : Questions sumQuestion 메서스 변경( 메서드명, 변수명)

메서드명은 sum 보다는 add 가 두 리스트를 더해서 반환한 다는 의미를 잘 표현한 거 같음

sumQuestions은 두 리스트를 더해서 반환한다는 의미를 가진다. unFixedQuestions 이라는 변수명은 비고정질문에 대한 의미로 한정 지어짐
따라서 좀 더 포괄적인 newQuestions이라는 새로운 질문이라는 의미의 변수로 변경

Jin409 commented on src/main/java/com/project/doongdoong/domain/question/model/Question.java:
```diff
@@ -29,28 +31,26 @@ public class Question extends BaseEntity {
     }
 
-    public void connectAnalysis(Analysis analysis){
+    public static Question of(QuestionContent questionContent) {
+        return new Question(questionContent);
+    }
```
인자가 하나인데 of 로 정적 팩토리 메서드를 사용하신 이유가 있나요?

Jin409 replied:
[링크](https://inpa.tistory.com/entry/GOF-%F0%9F%92%A0-%EC%A0%95%EC%A0%81-%ED%8C%A9%ED%86%A0%EB%A6%AC-%EB%A9%94%EC%84%9C%EB%93%9C-%EC%83%9D%EC%84%B1%EC%9E%90-%EB%8C%80%EC%8B%A0-%EC%82%AC%EC%9A%A9%ED%95%98%EC%9E%90#%EC%A0%95%EC%A0%81_%ED%8C%A9%ED%86%A0%EB%A6%AC_%EB%A9%94%EC%84%9C%EB%93%9C_%EB%84%A4%EC%9D%B4%EB%B0%8D_%EA%B7%9C%EC%B9%99)에 따르면 of 메서드를 사용하는 경우는 인자가 여러개일 때 적합한 것으로 보여서요!

Kimseungin0529 replied:
감사합니다

Jin409 commented on src/main/java/com/project/doongdoong/domain/question/model/QuestionContent.java:
```diff
@@ -3,15 +3,15 @@
 
     FIXED_QUESTION1("어떤 이유로 상담이 필요하신가요?", 1, true),
-    FIXED_QUESTION2("당신의 감정에 무엇이 가장 큰 영향을 주었나요?",2, true),
+    FIXED_QUESTION2("당신의 감정에 무엇이 가장 큰 영향을 주었나요?", 2, true),
```
DB 에 저장하지 않고 열거형으로 선언한 이유가 있나요?

Kimseungin0529 replied:
생각해보니 명확한 이유가 없었던 거 같아요. 어떤 방법이 더 나을까요?

Jin409 commented on src/main/java/com/project/doongdoong/domain/question/model/QuestionContent.java:
```diff
@@ -26,72 +26,27 @@ public enum QuestionContent {
+    private static final List<QuestionContent> UNFIXED_QUESTION_CONTENTS = List.of(
+            UNFIXED_QUESTION1, UNFIXED_QUESTION2, UNFIXED_QUESTION3, UNFIXED_QUESTION4
+    );
+    private static final Random random = new Random();
```
random 은 randomIndex 메서드에서만 사용되는 걸로 보이는데 필드로 선언한 이유가 있을까요?

Kimseungin0529 replied:
randomIndex 메서드가 호출될 때마다 random 객체를 사용해야 하는데 그때마다 객체를 생성하는 것보다 static 필드로 선언해서 하나의 객체를 사용하는 게 더 객체 생성을 방지할 수 있어서 좋아 보여서 했습니다.

Jin409 commented on src/main/java/com/project/doongdoong/domain/question/service/QuestionProvider.java:
```diff
@@ -0,0 +1,59 @@
+        return questionsCandidate.extractRandomQuestions(size);
+    }
+
+    private List<Question> getQuestionListFrom(List<QuestionContent> questionContents) {
```
복수의 Question 에 대해서 어떤 메서드에서는 `questions` 이고, 어떤 메서드에서는 `questionList` 인데 구분 기준이 있을까요?

Kimseungin0529 replied:
일급 컬렉션인 Quesions 적절한 필드명이 questions이라고 생각했어요.
그런데 Questions 의 메서드를 구현하다 보니 Questions의 필드인 List<Question>와 파라미터로 받는 List<Question>의 네이밍이 겹쳤습니다.

 Questions의 필드인 List<Question>을 quesions라 표현했고 파라미터로 받는 List<Question>를 리스트의 의미를 주기 위해 question, questionList로 구분했습니다.

좀 더 적절한게  있으면 조언해주십숑

Commit:
refactor : Question 정적 팩토리 메서드명 변경 of -> from

하나의 인자를 받기에 더 적절할 from으로 변경

Commit:
chore : Rest docs 기능 추가

Commit:
feat : RestDocs 사용을 위한 서포트 클래스 생성

Commit:
feat/chore : RestDocs 문서 구현 및 파일 추가

1. build.gradle.kts -> RestDocs 의존성 추가 및 경로 설정
2. RestDocs 관련 클래스 생성 및 테스트와 문서 연동
Secuirty 허용 경로 추가 및 RestDocs 커스터마이징 추가

Commit:
chore : 분석 생성, 분석 단일 조회 API 해피 케이시 문서 작성

Commit:
chore : 분석 페이징 조회, 분석 당일 기준 일주일 기록 조회, 분석 질문지에 대한 답변 저장, 분석 질문지 삭제 API 해피 케이스 문서 작성(RestDocs)
